Add 2D CSG boolean operations

smix8 opened the pull request:
Adds nodes and functions for 2D CSG boolean operations.

Implements proposal https://github.com/godotengine/godot-proposals/issues/3731

### PR to-do / status

<details>

<summary>Detail</summary>

#### Core
- [X] Create `csg_2d` engine module.
- [X] Add boolean operations and brush logic for union, intersection and subtraction.
- [X] Add CSG node types for rectangle, circle, capsule, mesh resource, polygon edit and combiner.
#### Brushes
- [X] Add CSGCapsule2D brush builder.
- [X] Add CSGCircle2D brush builder.
- [X] Add CSGMesh2D brush builder.
- [X] Add CSGPolygon2D brush builder.
- [X] Add CSGRectangle2D brush builder.
#### CSG result
- [X] Add triangle mesh creation.
- [X] Add option for mesh UV creation.
- [X] Add option for mesh vertex color creation.
#### Physics
- [X] Add CSG root node collision using concave segments shapes.
- [X] Add CSG root node collision using convex polygons shapes.
#### Conversions
- [X] Add CSG result to MeshInstance2D node conversion.
- [X] Add CSG result to CollisionShape2D nodes conversion.
- [X] Add CSG result to Polygon2D node conversion.
- [X] Add CSG result to LightOccluder2D nodes conversion.
#### Editor and Tooling
- [X] Add Editor EditorPluginCSG2D.
- [X] Add Editor CSGShape2DEditor.
- [X] Add Editor menu options for CSG root node.
- [x] Add Editor tooling for CSGCapsule2D.
- [x] Add Editor tooling for CSGCircle2D.
- [x] Add Editor tooling for CSGPolygon2D.
- [x] Add Editor tooling for CSGRectangle2D.
- [X] Add CSG icons for 2D.
#### Documentation
- [x] Add class documentation for CSGCapsule2D.
- [x] Add class documentation for CSGCircle2D.
- [X] Add class documentation for CSGCombiner2D.
- [x] Add class documentation for CSGMesh2D.
- [x] Add class documentation for CSGPolygon2D.
- [x] Add class documentation for CSGPrimitive2D.
- [x] Add class documentation for CSGRectangle2D.

#### Polish
- [x] Code cleanup and restructuring.
- [x] Bug sweep.
- [x] Update PR description.

</details>

### Wait, what is CSG?

![csg2d_bake_01](https://github.com/user-attachments/assets/0985e1b2-c20f-4e61-8ce9-3a3461c600bb)

To quote from the older 3D CSG blog that can be found [here](https://godotengine.org/article/godot-gets-csg-support/).

> CSG stands for “Construtive Solid Geometry”, and is a tool to combine basic (and not so basic) shapes to create more complex shapes. In the 3D modelling software, CSG is mostly known as “Boolean Operators”.

So basically you can do everything what you can do currently in scripts using the `Geometry2D` merge and slice related boolean operation functions but more convenient with nodes. There are also plenty of extra features on top.

### CSG Node types

Largely copies the workflow and nodes from the 3D CSG although accounting for some specific 2D quirks.

![csg2d](https://github.com/user-attachments/assets/674536ca-2e3c-40e3-84a2-708879561d92)

- CSGCapsule2D
Capsule shape with radius, height and corner segment properties.
- CSGCircle2D
Basic circle with radius and radial segment properties to control detail and turn the shape into e.g. a hexa or with only 3 segments into a triangle.
- CSGCombiner2D
Does nothing itself other than combining csg children and stops the csg propagation for better organisation.
- CSGMesh2D
Accepts a rendering mesh resource that uses the `ARRAY_FLAG_USE_2D_VERTICES` format.
- CSGPolygon2D
Drawable polygon outline similar to e.g. Polygon2D node.
- CSGRectangle2D
Basic rectangle with size property.

Each CSG root node can optionally `use_collision`. Depending on the `collision_shape_type` property this will be either a single concave collision shape or multiple convex collision shapes.

Note that certain Editor toolbar menu bake options (see further below) require that collision is enabled. Without collision enabled some of the very expensive shape data is not created so it can not be exported.

### CSG Operators

The available operators are the same as in 3D:
- UNION
- INTERSECTION
- SUBTRACTION

Union merges the shape to the parent or higher sibling node shape while substraction cuts into them. The intersection is the weird one that removes all shape parts that are not found in the combined shapes.

Operations are done in node tree order same as in 3D, or to quote from the old CSG 3D blog.

> Every CSG node will first process it’s children nodes (an their operation: union, intersection substraction), in tree order and apply them to itself one after the other.

Note that as with 3D the update of CSG shapes happens deferred. This is nessary because CSG results depend on other CSG nodes fully loaded and fully updated in the order for the final result. You can not spawn a CSG node and expect an immediate result. Either you wait for the nodes or you use the Geometry2D class for procedual stuff that has no node dependency. See related proposal https://github.com/godotengine/godot-proposals/issues/10395 for improving the usability with scrips.

### CSG root node and result

Only the CSG `root` note creates a result and has properties to customize it.

A CSG `root` node is either a `CSGCombiner2D` or any CSG Node that has no other CSG node as parent.

By default the result creates an indexed 2d triangle mesh for rendering visuals.
This rendering mesh has UV range mapped around the mesh `Rect`.

The option to add vertex colors to the mesh exists by enabling `use_vertex_color` property and picking the desired `vertex_color`.

### Debug

Inside the Editor the CSG result will color the mesh faces as well as display the edges of the convex polygons. This is not rendered outside the editor where the mesh is just plain white by default.

With a lot of CSG brushes involved things can get confusing what shape and brush does what.

 There is the option to display the CSG brush outline as debug colored by the operation type.
- OPERATION_UNION -> Green
- OPERATION_INTERSECTION -> Orange
- OPERATION_SUBTRACTION -> Red

This option can be enabled per CSG node individually with the `debug_show_brush` property.
By default this debug is off to not clutter the editor view so much.

![brush_debug](https://github.com/user-attachments/assets/58135761-2633-42d7-8fcf-a390bb5ea10a)

### Baking CSG results to static geometry

![csg_bake_options](https://github.com/user-attachments/assets/0b47643e-dbb9-4f39-b257-127ae97fc73f)

Similar to 3D PR https://github.com/godotengine/godot/pull/93252

CSG options to bake the CSG root node result to a static mesh or collision shapes or other node types. This can be used to "design" a level or some shape geometry with CSG and then bake the result to a more efficient static version for performance (or to avoid seam issues).

#### Bake to MeshInstance2D
Creates a MeshInstance2D node with a 2d triangle `ArrayMesh` that has `UV` mapped same as mesh `Rect` and optional vertex colors if  `use_vertex_color` is enabled.

#### Bake to CollisionShape2D
Creates multiple CollisionShape2D nodes with either convex polygon shapes or concave segments shapes depending on `collision_shape_type`.

#### Bake Polygon2D
Creates a Polygon2D node and adds the CSG vertices to the `polygon` array and the CSG convex polygons indices to the `polygons` array.

Note that the Editor tooling of Polygon2D is broken with multiple polygons. This is an issue with Polygon2D editor plugin and not with the CSG conversion.

#### Bake LightOccluder2D
Creates multiple LightOccluder2D nodes from the CSG outlines, The used OccluderPolygon2D resources are not closed as that would not work if there are any holes in the CSG shapes. So if you want some of the occluders closed you need to set that manually.

#### Bake NavigationRegion2D
Creates a NavigationRegion2D with a NavigationPolygon resource and sets the vertices and convex polygon indices same as the CSG result.

#### Bake with scripts

There is also the option to create all the resources in script, not creating or involving additional nodes.

```gdscript
var baked_mesh: ArrayMesh = CSGShape2D.bake_static_mesh()
# Visual 2d mesh with UV based on Rect size and optional vertex colors.

var baked_shapes: Array = CSGShape2D.bake_collision_shapes()
# 2d has no sensible single shape resource that works for more complex shapes
# so it needs to return an array that has multiple convex or concave shapes.

var baked_navigation_mesh: NavigationPolygon = CSGShape2D.bake_navigation_mesh()

var baked_light_occluders: Array = CSGShape2D.bake_light_occluders()
# An array of multiple OccluderPolygon2D.
```

### Why name it CSG in 2D?
Although CSG "Construtive Solid Geometry” is more a name used in 3D modelling context I stayed with the name for 2D because users are already very familiar with the term in Godot from 3D. The 2D and 3D nodes and workflows are kept very similar on purpose as it allows better knowledge and documentation sharing.

### Performance

Compared to the far more complex 3D CSG the 2D version has actually pretty good performance for what it does. Although I still would not recommend planning to use it with hundreds of changing sub nodes at runtime.

For rendering performance, if only the CSG root node transform is changed that costs basically nothing at runtime. The 2D CSG does not use the Node2D draw functions for the polygons or lines like many other 2d related nodes. The actual rendering geometry is baked to a single static 2d mesh in the end of the operations. So the entire performance cost of moving the CSG root node at runtime is a `canvas_item_set_transform()` call.

This is similar to what the navmesh baking does but the huge difference for runtime change performance is that the specialised CSG nodes can all catch their own intermediate result. So on changes only the parts that actually change up in the tree order need to be reparsed and recalculated instead of absolutely everything.

### Help! My polygons are all breaking!

The boolean operations are done with the Clipper2 polytree backend base on polygon outline paths. As such all the usual polygon outline limitations apply that can be read in detail here https://angusj.com/clipper2/Docs/Robustness.htm.

These outlines need to be converted to either triangles or convex polygons in the end. Any kind of resulting overlap or crossed edges, e.g. due to float precision issues, can break those conversions so work with some margin in mind.

As always when dealing with outline to polygon conversions, dont (upscale) float precision fumble your layouts, avoid self-intersection at all time and **never cross the (edge) streams** in any way guys!

![crossthestreams](https://github.com/user-attachments/assets/040a2611-2247-438e-9e42-05b5a5532865)
_Don't cross the (edge) streams!_

The CSG can fix a lot of weird shapes due to various merge steps but if the source geometry has already grievous geometry errors the CSG chain still can break. This will be more a problem with the CSGPolygon2D and CSGMesh2D nodes as they allow the creation of all kinds of invalid geometry input. It can also happen when weird node scaling is used as this may cause vertices to end up in unintended rasterization cells when float positions are upscales back and forth. Same can happen when shapes that are perfectly aligned with shared vertices at corners in the editor. These kind of "pixel-perfectionist" layouts regularly stop to work the moment the float positions get upscaled as suddenly vertices may end up inside other shapes, either keep some error margin or create those shapes separated. You have been warned :)

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/3731*

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

fire commented:
Awesome, need some time to test and fix up the integration tests but looks promising.

smix8 committed:
Add CSG2D for boolean operations

Adds nodes and functions for 2D CSG boolean operations.

Calinou reviewed (commented):
Tested locally, it works as expected.

**Testing project:** [test_csg_2d.zip](https://github.com/user-attachments/files/18082748/test_csg_2d.zip)

![image](https://github.com/user-attachments/assets/7f71ab3b-8a30-49f7-8597-6225e51368da)

Some feedback:

- When **Use Vertex Color** is disabled, the CSG node will have a different appearance in the editor compared to the running project:

Editor | Running project
-|-
![image](https://github.com/user-attachments/assets/68dd43ab-348a-4ac8-b0c2-d15ed314a423) | ![image](https://github.com/user-attachments/assets/4661131a-5547-46e5-9825-ab2762db2ab0)

When enabled, it has the same appearance in the editor and running project. I'm guessing this may occur because when **Use Vertex Color** is disabled, the shape is still drawn in white but is drawn *behind* the debug gizmos.

- When **Use Vertex Color** is disabled, the Vertex Color property should be hidden in the inspector.

smix8 commented:
> When enabled, it has the same appearance in the editor and running project. I'm guessing this may occur because when **Use Vertex Color** is disabled, the shape is still drawn in white but is drawn _behind_ the debug gizmos.

That half transparent blue hue face color is editor specific to not have white solid shapes in the editor. It swaps that color on the mesh in the ` CSGShape2D::draw_shape()` function based on the editor hint. Same with the inner polygon edge line debug that is not added at runtime.

Calinou commented on modules/csg_2d/doc_classes/CSGShape2D.xml:
```diff
@@ -0,0 +1,118 @@
+	</brief_description>
+	<description>
+		The 2D CSG base class that provides CSG operation support to the various 2D CSG nodes in Godot.
+		[b]Performance:[/b] CSG nodes are only intended for prototyping as they have a significant CPU performance cost.
+		Consider baking final CSG operation results into static geometry that replaces the CSG nodes.
```
```suggestion
		[b]Performance:[/b] CSG nodes are only intended for prototyping as they have a significant CPU performance cost. Consider baking final CSG operation results into static geometry that replaces the CSG nodes.
```

Calinou commented on modules/csg_2d/doc_classes/CSGMesh2D.xml:
```diff
@@ -0,0 +1,18 @@
+		A 2D CSG mesh shape.
+	</brief_description>
+	<description>
+		A 2D CSG node that allows to create a shape for the 2D CSG system based on the geometry of a 2d rendering mesh resource.
```
```suggestion
		A 2D CSG node that allows to create a shape for the 2D CSG system based on the geometry of a 2D rendering mesh resource.
```

smix8 commented:
Added some additional debug visuals. It was getting confusing with more CSG brushes involved what brush does what.

 There is now the option to display the CSG brush outline as debug colored by the operation type.
- OPERATION_UNION -> Green
- OPERATION_INTERSECTION -> Orange
- OPERATION_SUBTRACTION -> Red

This option can be enabled per CSG node individually.
By default this debug is off to not clutter the editor view so much.

![brush_debug](https://github.com/user-attachments/assets/c670a283-acfc-428a-8472-92eecf782d8d)

smix8 commented:
Pushed the last addition / changes to the PR. I think it is now good for review.

smix8 commented:
Added the option to bake the CSG result to a navigation mesh with `bake_navigation_mesh()`.
Added editor menu option to create a NavigationRegion2D with that baked navmesh.

smix8 commented:
Added a full fledged polygon editor with toolbar for the CSGPolygon2D.

smix8 commented:
Fixed a bug where the CSG would not react when the child node order changes.
This code was a direct copy from the 3D CSG that seem to have the same problem.

Mickeon reviewed (commented):
Not much to say on the documentation side. It is actually all borrowed from the 3D equivalents. Its quality varies, but it's better to be consistent here (for the translators' sanity), and adjustments can be made on both another time.

rburing commented on modules/csg_2d/csg_shape_2d.cpp:
```diff
@@ -0,0 +1,1046 @@
+		root_collision_debug_instance = RS::get_singleton()->instance_create();
+	}
+
+	//RenderingServer::get_singleton()->canvas_item_add_mesh(get_canvas_item(),  root_edge_line_mesh->get_rid(), Transform2D(), color, texture_rid);
```
I would add a `TODO` or a `FIXME` comment here, if it's not implemented in this PR.

smix8 replied:
Those collision debug related functions are not really doing something. They exists to stay parallel with the CSG3D in case someone wants to do anything with it. I did not add special collision debug here because it would just create overlap noise in 2D as the default csg 2d visuals already show the edges compared to 3D.

rburing commented on modules/csg_2d/csg_shape_2d.cpp:
```diff
@@ -0,0 +1,1046 @@
+void CSGShape2D::_on_transform_changed() {
+	if (root_collision_debug_instance.is_valid() && !debug_shape_old_transform.is_equal_approx(get_global_transform())) {
+		debug_shape_old_transform = get_global_transform();
+		//RS::get_singleton()->instance_set_transform(root_collision_debug_instance, debug_shape_old_transform);
```
Same here. (Also this `instance` naming is a bit 3D-ish, but no big deal.)

rburing commented on modules/csg_2d/csg_shape_2d.cpp:
```diff
@@ -0,0 +1,1046 @@
+	if (!is_root_shape()) {
+		return;
+	}
+	//CSGBrush2D *n = _get_brush();
```
Can be removed.

rburing commented on modules/csg_2d/doc_classes/CSGPrimitive2D.xml:
```diff
@@ -0,0 +1,12 @@
+		Base class for 2D CSG primitives.
+	</brief_description>
+	<description>
+		Parent class for various 2D CSG primitives. It contains code and functionality that is common between them. It cannot be used directly. Instead use one of the various classes that inherit from it.
```
This should say base class like in the brief description.

rburing commented on modules/csg_2d/editor/csg_editor_plugin_2d.cpp:
```diff
@@ -0,0 +1,804 @@
+		} break;
+
+		case CSGShapeType::MESH_SHAPE: {
+			// CSGMesh2D *shape = Object::cast_to<CSGMesh2D>(node);
```
Add a `FIXME` or `TODO` comment?

smix8 replied:
There is not really anything to do here as that `CSGShapeType::MESH_SHAPE` has no editor overlay.

rburing reviewed (changes requested):
This mostly looks great to me! I have just some small comments.

smix8 commented on modules/csg_2d/csg_capsule_2d.h:
```diff
@@ -0,0 +1,71 @@
+	float get_height() const;
+
+	void set_corner_segments(const int p_corner_segments);
+	int get_corner_segments() const;
```
renamed

smix8 commented on modules/csg_2d/csg_capsule_2d.cpp:
```diff
@@ -0,0 +1,147 @@
+	if (vertices_cache_dirty) {
+		vertices_cache_dirty = false;
+
+		vertices_cache.resize(radial_segments * 4 + 2);
```
How about I conjure a quarter circle documentation.

KoBeWi commented:
Baking CollisionShape fails for simple case:
![image](https://github.com/user-attachments/assets/03bff50b-f5c3-47cd-8d9a-4e92b32f0f7d)
Either it should create ConcavePolygonShape2D here, or you could add an option for CollisionPolygon2D (or the option could automatically fallback to CollisionPolygon2D if shape fails).

The bigger rectangle for rectangle shape looks weird:
![image](https://github.com/user-attachments/assets/328dbf2e-5ad8-400b-9211-cce1c50051fe)
It makes sense for other shapes, because it allows to make e.g. ovals, but for rectangle it's redundant. They don't need to be scaled I think.

KoBeWi commented on modules/csg_2d/csg_capsule_2d.cpp:
```diff
@@ -0,0 +1,147 @@
+
+#include "thirdparty/misc/polypartition.h"
+
+#ifdef DEBUG_ENABLED
```
Should be TOOLS_ENABLED.

KoBeWi replied:
Sorry, I was wrong. I missed when this method was moved to `DEBUG_ENABLED` due to same Game tab stuff.

KoBeWi commented on modules/csg_2d/csg_capsule_2d.h:
```diff
@@ -0,0 +1,71 @@
+	void set_radial_segments(const int p_segments);
+	int get_radial_segments() const;
+
+	CSGCapsule2D() {}
```
```suggestion
```
Empty constructors are unnecessary.

KoBeWi commented on modules/csg_2d/csg_2d.cpp:
```diff
@@ -0,0 +1,86 @@
+	Clipper2Lib::ClipperD clipper_D;
+	clipper_D.PreserveCollinear(false);
+
+	for (uint32_t i = 0; i < p_outlines.size(); i++) {
```
```suggestion
	for (const Outline &outline : p_outlines) {
```

KoBeWi commented on modules/csg_2d/csg_shape_2d.h:
```diff
@@ -0,0 +1,216 @@
+#include "csg_2d.h"
+
+#include "scene/2d/node_2d.h"
+#include "scene/resources/2d/concave_polygon_shape_2d.h"
+#include "scene/resources/2d/convex_polygon_shape_2d.h"
+#include "scene/resources/2d/navigation_polygon.h"
```
These can be forward-declared.

Calinou reviewed (commented):
Tested locally, it works as expected.

Some feedback:

- **Bake CollisionShape2D** should be renamed to **Bake CollisionShape2D(s)**, as it can create multiple collision shapes in a single operation if the collision shape is convex but the combiner's shape is concave. When the collision shape is concave, it will also occur if the CSG combiner has multiple disjoint shapes in it. This does not occur with other baking types (even Polygon2D, surprisingly).
- **Bake Polygon2D** does not reuse the vertex color that was defined in CSGCombiner2D, so it resets to white. In comparison, the color is preserved when baking to MeshInstance2D.

smix8 commented:
> This does not occur with other baking types ...

There are other bake types that can result into multiple shapes or nodes depending on settings and composition, e.g. the light occluders. I added them but frankly those (s) behind node names look visually poor, especially on buttons.

> Bake Polygon2D does not reuse the vertex color ...

Added that if `use_vertex_color` is enabled on the CSG that the baked Poygon2D has that color copied into the `vertex_colors` property.

AThousandShips commented on modules/csg_2d/csg_shape_2d.cpp:
```diff
@@ -0,0 +1,1046 @@
+	TPPLPartition tpart;
+
+	if (tpart.ConvexPartition_HM(&tppl_in_polygon, &tppl_out_polygon) == 0) {
+		ERR_PRINT("CSGShape2D outines to convex polygons conversion failed. Check input geometry for common shape outline and precision errors like self-overlapping shapes used in intersection operations.");
```
To confirm, this isn't a critical situation, it should continue here after failing?

smix8 replied:
It needs to run the code past that point for e.g. redraw so that the visuals actually update with the failed state.

AThousandShips commented on modules/csg_2d/doc_classes/CSGCapsule2D.xml:
```diff
@@ -0,0 +1,23 @@
+		A 2D CSG Capsule shape.
+	</brief_description>
+	<description>
+		This node allows to create a capsule for use with the 2D CSG system.
```
```suggestion
		This node allows creating a capsule for use with the 2D CSG system.
```

AThousandShips commented on modules/csg_2d/doc_classes/CSGCircle2D.xml:
```diff
@@ -0,0 +1,20 @@
+		A 2D CSG circle shape.
+	</brief_description>
+	<description>
+		A 2D CSG node that allows to create circle shapes or other regular convex polygons for use with the 2D CSG system.
```
```suggestion
		A 2D CSG node that allows creating circle shapes or other regular convex polygons for use with the 2D CSG system.
```

AThousandShips commented on modules/csg_2d/doc_classes/CSGCombiner2D.xml:
```diff
@@ -0,0 +1,12 @@
+		A 2D CSG node to structure other 2D CSG nodes.
+	</brief_description>
+	<description>
+		For complex arrangements of shapes, it is sometimes needed to add structure to CSG nodes. A CSGCombiner2D node allows to create this structure. The node encapsulates the result of the CSG operations of its children. In this way, it is possible to do operations on one set of shapes that are children of one CSGCombiner2D node, and a set of separate operations on a second set of shapes that are ch ...
```
```suggestion
		For complex arrangements of shapes, it is sometimes needed to add structure to CSG nodes. A CSGCombiner2D node allows creating this structure. The node encapsulates the result of the CSG operations of its children. In this way, it is possible to do operations on one set of shapes that are children of one CSGCombiner2D node, and a set of separate operations on a second set of shapes that are children of a second CSGCombiner2D node, and then do an operation that takes the two end results as its input to create the final shape.
```

AThousandShips commented on modules/csg_2d/doc_classes/CSGMesh2D.xml:
```diff
@@ -0,0 +1,17 @@
+		A 2D CSG mesh shape.
+	</brief_description>
+	<description>
+		A 2D CSG node that allows to create a shape for the 2D CSG system based on the geometry of a 2D rendering mesh resource.
```
```suggestion
		A 2D CSG node that allows creating a shape for the 2D CSG system based on the geometry of a 2D rendering mesh resource.
```

AThousandShips commented on modules/csg_2d/doc_classes/CSGPolygon2D.xml:
```diff
@@ -0,0 +1,17 @@
+		A 2D CSG polygon shape.
+	</brief_description>
+	<description>
+		A 2D CSG node that allows to create a polygon for use with the 2D CSG system. For CSG union and intersection operations the polygon can be arbitrarily shaped. For CSG subtraction operations the polygons must be non-self-overlapping.
```
```suggestion
		A 2D CSG node that allows creating a polygon for use with the 2D CSG system. For CSG union and intersection operations the polygon can be arbitrarily shaped. For CSG subtraction operations the polygons must be non-self-overlapping.
```

AThousandShips commented on modules/csg_2d/doc_classes/CSGRectangle2D.xml:
```diff
@@ -0,0 +1,17 @@
+		A 2D CSG rectangle shape.
+	</brief_description>
+	<description>
+		A 2D CSG node that allows to create a rectangle for use with the 2D CSG system.
```
```suggestion
		A 2D CSG node that allows creating a rectangle for use with the 2D CSG system.
```

AThousandShips reviewed (commented):
Will do a second more in-depth pass over the documentation as well

KoBeWi commented:
From the things I reported previously, only the polygon color was fixed.
You should fix at least debug brushes, because they are completely broken with undo and duplication.

You could also resolve my point about failed CollisionShape2D baking. Polygon2D is already an option, so fallback to CollisionPolygon2D is possible too. CollisionShape2D and PolygonShape2D are more or less interchangeable and you can use the same data as in Polygon2D.
![image](https://github.com/user-attachments/assets/4a425665-f22b-4d72-94db-ebdda643c338)

> I added it for consistency with other 2d toolings. In general I do not like scaling of 2d nodes

Then you don't need to add it ;)
Especially for CSGRectangle2D it's almost useless and looks ugly. Scaling is still possible without gizmos with the Scale Tool.

smix8 commented:
I rebased to not be 3500++ PRs behind master (time flies).

- Addressed the documentation comments from ship (I think?)
- Removed the Node2D related scaling boxes.

smix8 commented on modules/csg_2d/csg_shape_2d.h:
```diff
@@ -0,0 +1,216 @@
+
+#include "csg_2d.h"
+
+#include "scene/2d/node_2d.h"
```
Removing this include breaks the build.

KoBeWi commented:
> The fallback to CollisionPolygon2D is imo not a solution to do by default because that polygon is just a visual outline that also need to convert to convex polygons or segments when used as collision. If a user sets it to convex it will break either way when the input breaks it as it uses the same convex decomp behind the scene as the csg.

Right now there is no way to make CollisionPolygon2D at all. You need to make Polygon2D and then change type, afterwards the polygon works correctly. I don't see why this can't be an option. The reason conversion "fails" is because CollisionShape2D is very limited. You can't bake e.g. a triangle, which is a perfectly valid shape.
![image](https://github.com/user-attachments/assets/7e39c3f0-b25a-4842-bc8a-b56582b42fca)

KoBeWi commented on modules/csg_2d/csg_shape_2d.cpp:
```diff
@@ -0,0 +1,1045 @@
+}
+
+void CSGShape2D::force_shape_update() {
+	shape_update_forced = true;
```
This seems unused.

smix8 replied:
removed

smix8 commented:
@KoBeWi silly question but did you forget to enable `use_collision` in your tests on the root node?
> Note that certain Editor toolbar menu bake options (see further below) require that collision is enabled. Without collision enabled some of the very expensive shape data is not created so it can not be exported.

Because without `use_collision` enabled I get the very same bake errors as you with that same node and shape setup. With collision enabled it all works just fine for me.

akien-mga commented:
Moving this to the 4.6 milestone as we're entering feature freeze.
Implementation wise this PR seems mostly ready so it could nearly have made it into 4.5 before freeze, but after discussion we think that we need some more time to evaluate how needed this feature is in the first place.

The proposal it implements got a total of 4 upvotes in 4 years, which isn't much. This PR has significantly higher success but it also just sounds cool :sunglasses: which doesn't necessarily correlate an actual need for users.

We should also evaluate this in context of other related proposals such as https://github.com/godotengine/godot-proposals/issues/1126 (which on the other hand has very high demand) and #100574 and #99210 which implement it (and conversely have seen less organic engagement than this PR).

All this to say, this might very well be the right tool for the job, but let's make sure this is properly assessed and solves the needs from the community.

smix8 commented:
> This needs to be mentioned somewhere. Also collision baking should be disabled when `use_collision` is disabled, if it does nothing.

Changed it so that when no csg collision exists or is enabled that the editor menu option and function create all the collision shapes procedural.

KoBeWi commented:
The `debug_show_brush` bug I mentioned earlier still exists, and there is a similar bug where if you duplicate a shape under CSGCombiner, the CSG menu will appear while it shouldn't. Though I think this can be ironed out later, all other problems seem to be resolved.

KoBeWi commented on modules/csg_2d/editor/csg_editor_plugin_2d.cpp:
```diff
@@ -0,0 +1,810 @@
+	options->set_text(TTR("CSG2D"));
+	options->set_switch_on_hover(true);
+
+	options->get_popup()->add_item(TTR("Bake MeshInstance2D"), MENU_OPTION_BAKE_MESH_INSTANCE);
+	options->get_popup()->add_item(TTR("Bake CollisionShape2D(s)"), MENU_OPTION_BAKE_COLLISION_SHAPE);
+	options->get_popup()->add_item(TTR("Bake Polygon2D"), MENU_OPTION_BAKE_POLYGON_2D);
+	options->get_popup()->add_item(TTR("Bake LightOccluder2D(s)"), MENU_OPTION_BAKE_LIGHT_OCCLUDER_2D);
+	options->get_popup()->add_item(TTR("Bake NavigationRegion2D"), MENU_OPTION_BAKE_NAVIGATION_REGION_2D);
```
```suggestion
	options->get_popup()->add_item(TTRC("Bake MeshInstance2D"), MENU_OPTION_BAKE_MESH_INSTANCE);
	options->get_popup()->add_item(TTRC("Bake CollisionShape2D(s)"), MENU_OPTION_BAKE_COLLISION_SHAPE);
	options->get_popup()->add_item(TTRC("Bake Polygon2D"), MENU_OPTION_BAKE_POLYGON_2D);
	options->get_popup()->add_item(TTRC("Bake LightOccluder2D(s)"), MENU_OPTION_BAKE_LIGHT_OCCLUDER_2D);
	options->get_popup()->add_item(TTRC("Bake NavigationRegion2D"), MENU_OPTION_BAKE_NAVIGATION_REGION_2D);
```

kitbdev commented:
The proposal is lacking detail and mentions destructive terrain as a use case, but that was already called out as not being a good fit. So the use case for this seems to be prototyping, just like 3D CSG.
But I don't think it is really needed in 2D.
I feel like 2D CSG isn't as popular as 3D CSG since there are other easy ways to make 2D shapes. We have Polygon2D which can already be used to create custom shapes. Also sprites can be used instead in many cases.

For basic shapes, https://github.com/godotengine/godot-proposals/issues/1126, I think either https://github.com/godotengine/godot/pull/100574  or https://github.com/godotengine/godot/pull/99210 are a better fit. Those approaches are simpler, easier to use, and don't have the performance considerations that this does.
For more complex shapes, you can use multiple shapes together and Polygon2D.
If the user needs it to be a single combined 2D mesh and Polygon2D won't work for some reason, then it's more niche or not really a prototype anymore.

I'm sure 2D CSG is good for some workflows, so I think 2D CSG would be a better fit as an addon. As mentioned, the base functionality for it is already available through Geometry2D, and it is very self-contained.

shafnaz commented:
Is this going to be available in 4.6?

rburing reviewed (approved):
As mentioned above there are some orthogonal use cases and solutions in this context. Personally I think 2D CSG is great for prototyping and constructing more complex shapes and then baking them. Being node-based this empowers code-challenged users to conveniently create and modify interesting shapes with collisions enabled. I think eventually more multi-shape operations should be added to Geometry2D or some other low-level interface exposed to scripting (wrapping Clipper2) and 2D CSG should be calling them (instead of using Clipper2 directly), but that can be done later.
